### DISCUSSION TITLE
MAUI

### DISCUSSION THREAD

**Discussion by IanBUK** [dotnet-bluetooth-le/dotnet-bluetooth-le#707] [Q&A] [answered]
Will this library be ported to work with MAUI, as Xamarin Forms are on their way out?

**Comment by janusw** [comment, marked as answer]
The current beta release 3.0.0-beta.4 supports .NET 6 and 7 and should work well with MAUI. Does this answer your question?

**Reply by IanBUK** [reply]
We'll find out shortly!

I'll update later this afternoon.

**Reply by IanBUK** [reply]
OK, so taking the code from:

`https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/releases/tag/3.0.0-beta.4`


Downloading it into a new folder and opening in Visual Studio for Mac 17.6, build 1575, Selecting "unload project" for the iOS, macOS andUWP projects, and then building gives the following error:

`/Applications/Visual Studio.app/Contents/MonoBundle/MSBuild/Current/bin/Microsoft.Common.CurrentVersion.targets(5,5): Error: Project '../../MvvmCross.Plugins.BLE/MvvmCross.Plugins.BLE.csproj' targets 'net6.0-android;net6.0-ios;net6.0-maccatalyst;net7.0-android;net7.0-ios;net7.0-maccatalyst'. It cannot be referenced by a project that targets 'MonoAndroid,Version=v12.0'. (BLE.Client.Droid)`

If I change the project settings to clear "Build with MSBuild on Mono", I get the following:

```
Getting restore information for project /Users/ianbrown/Documents/GitHubRepos/dotnet-bluetooth-le-3.0.0-beta.4/Plugin.BLE.Tests/Plugin.BLE.Tests.csproj
Getting restore information for project /Users/ianbrown/Documents/GitHubRepos/dotnet-bluetooth-le-3.0.0-beta.4/BLE.Client/BLE.Client.Droid/BLE.Client.Droid.csproj
/Users/ianbrown/Documents/GitHubRepos/dotnet-bluetooth-le-3.0.0-beta.4/BLE.Client/BLE.Client.Droid/BLE.Client.Droid.csproj(142,3) : error MSB4019: The imported project "/usr/local/share/dotnet/sdk/7.0.305/Xamarin/Android/Xamarin.Android.CSharp.targets" was not found. Confirm that the expression in the Import declaration "/usr/local/share/dotnet/sdk/7.0.305//Xamarin/Android/Xamarin.Android.CSharp.targets" is correct, and that the file exists on disk.
Unable to create package spec for project. '/Users/ianbrown/Documents/GitHubRepos/dotnet-bluetooth-le-3.0.0-beta.4/BLE.Client/BLE.Client.Droid/BLE.Client.Droid.csproj'

```

If, in Visual Studio I open the SDK Command Prompt and type `dotnet --info` I get the following:

```
.NET SDK:
 Version:   7.0.305
 Commit:    98e1b6c381

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  13.4
 OS Platform: Darwin
 RID:         osx.13-arm64
 Base Path:   /usr/local/share/dotnet/sdk/7.0.305/

Host:
  Version:      7.0.8
  Architecture: arm64
  Commit:       4b0550942d

.NET SDKs installed:
  6.0.300 [/usr/local/share/dotnet/sdk]
  6.0.301 [/usr/local/share/dotnet/sdk]
  6.0.302 [/usr/local/share/dotnet/sdk]
  6.0.406 [/usr/local/share/dotnet/sdk]
  6.0.410 [/usr/local/share/dotnet/sdk]
  6.0.411 [/usr/local/share/dotnet/sdk]
  7.0.200 [/usr/local/share/dotnet/sdk]
  7.0.203 [/usr/local/share/dotnet/sdk]
  7.0.304 [/usr/local/share/dotnet/sdk]
  7.0.305 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.7 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.14 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.18 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.19 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.7 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.8 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.6 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.7 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.14 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.18 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.19 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.7 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.8 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  x64   [/usr/local/share/dotnet/x64]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

None of the Xamaring.[OSPLATFORM].CSharp.targets files are on my machine, yet VS tells me that all is present and correct.

**Reply by janusw** [reply]
If you want to build on Mac, you should use `BLE.Mac.slnf` instead of `BLE.sln`.

**Comment by IanBUK** [comment]
My fork, at https://github.com/IanBUK/dotnet-bluetooth-le, now has a Maui sample client. It's a very cut down version from the ones you provide. But...it build on macOS in VS2023, and runs, and can scan BLE devices. You have to ensure all permissions and Location permissions are enabled.

**Reply by janusw** [reply]
Sounds great. Would you be willing to open a PR from it? It might be a good start for a MAUI sample client.

**Reply by IanBUK** [reply]
Yes, I’ll do that next time I get chance.

**Reply by IanBUK** [reply]
I got my Maui client to build and run, my code needs a little tidying, so I'll do that first.

However, my main need is to get this running under Godot 3.x. I've finally got it to all build - your build scripts are more complex than I was expecting ... - but I get an error when referencing from Godot. I finally got it all to compile by reverting to the nuget of your library, based on version 2.1. Basically, it's now complaining that it can't find Mono.Android version 0.0.0.0, even when I copy Mono.Android.DLL into place.

I'm assuming this is some sort of binding issue and a redirect might be in order. 

Sadly, I only get a few hours a week on this, so progress is slow.
